File: worker.cc
#include "zmq_util.h"

#include <gflags/gflags.h>
#include <glog/logging.h>
#include <zmq.hpp>

#include <exception>
#include <iostream>
#include <string>

DEFINE_string(in, "tcp://localhost:8001", "Frontend socket");
DEFINE_string(out, "tcp://localhost:8002", "Backend socket");
DEFINE_int32(io_threads, 1, "Number of threads dedicated to I/O operations");

using namespace std;

int main(int argc, char* argv[]) {
  google::ParseCommandLineFlags(&argc, &argv, true);
  google::InitGoogleLogging(argv[0]);

  zmq::context_t context(FLAGS_io_threads);
  zmq::socket_t fe(context, ZMQ_PULL);
  zmq::socket_t be(context, ZMQ_PUSH);

  try {
    fe.connect(FLAGS_in.c_str());
  } catch (std::exception& e) {
    LOG(FATAL) << "Failed to connect frontend socket: " << FLAGS_in << std::endl;
  }

  try {
    be.connect(FLAGS_out.c_str());
  } catch (std::exception& e) {
    LOG(FATAL) << "Failed to connect backend socket: " << FLAGS_out << std::endl;
  }

  while (true) {
    string uid = s_recv(fe);
    string req = s_recv(fe);
    cout << "Received: " << req << endl;

    string rep = "done_" + req;
    s_send(be, rep);
  }
}

File: worker.py
#! /usr/bin/env python
import argparse
import sys
import zmq

if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--in', default = 'tcp://localhost:8001', help = 'Incoming socket')
    parser.add_argument('--out', default = 'tcp://localhost:8002', help = 'Outgoing socket')
    options = vars(parser.parse_args())

    context = zmq.Context()
    fe = context.socket(zmq.PULL)
    be = context.socket(zmq.PUSH)

    try:
        fe.connect(options['in'])
    except:
        sys.stderr.write('Failed to connect incoming socket: %s\n' % options['in'])
        sys.exit(1)

    try:
        be.connect(options['out'])
    except:
        sys.stderr.write('Failed to connect outgoing socket: %s\n' % options['out'])
        sys.exit(1)

    while True:
        uid = fe.recv()
        req = fe.recv()
        print 'Received:', req

        rep = 'done_' + req
        be.send(rep)
        
